marlon/baseline_models/env_wrappers/defend_wrapper.py

```diff
@@ -0,0 +1,259 @@
+                    firewall_list.append(rule.port)
+            else:
+                for rule in node_info.firewall.incoming:
+                    firewall_list.append(rule.port)
+
+            if self.firewall_rule_list[port_from_action] not in firewall_list:
+                return False
+            return True
+
+        def service_exists(node_info: model.NodeInfo, service_from_action: int):
+            """Checks if a service exists on a node"""
+            # TODO IS THIS GOOD? This is just checking if its out of bounds...
+            # The names could be in a different order, so action 1 could be one service on one node and a differnet on another...
+            if service_from_action >= len(node_info.services):
+                return False
+            return True
+
+        if action[0] == 0:
+            # REIMAGE
+            _, node_info = get_node_and_info(action[1])
+            # If the node does not exist, or if the node is not currently running, this action is invalid.
+            return node_exists_and_running(action[1]) and is_reimagable(node_info)
+
+        elif action[0] == 1:
+            # block traffic
+            _, node_info = get_node_and_info(action[2])
+            # If the node does not exist, or if the node is not currently running, this action is invalid.
+            # The firewall rule needs to exist as well to block the traffic.
+            return node_exists_and_running(action[2]) and firewall_rule_exists(node_info, action[3], bool(action[4]))
+
+        elif action[0] == 2:
+            # allow traffic
+            _, node_info = get_node_and_info(action[5])
+            # If the node does not exist, or if the node is not currently running, this action is invalid.
+            return node_exists_and_running(action[5])
+
+        elif action[0] == 3:
+            # stop service
+            # If the node does not exist, or if the node is not currently running, this action is invalid.
+            # Also if the service to stop does not exist, this is invalid
+            _, node_info = get_node_and_info(action[8])
+            return node_exists_and_running(action[8]) and service_exists(node_info, action[9])
+
+        elif action[0] == 4:
+            # start service
+            # If the node does not exist, or if the node is not currently running, this action is invalid.
+            # Also if the service to start does not exist, this is invalid
+            _, node_info = get_node_and_info(action[10])
+            return node_exists_and_running(action[10]) and service_exists(node_info, action[11])
+        else:
+            return False
+
+    def reset(self) -> Observation:
+        self.valid_action_count = 0
+        self.invalid_action_count = 0
+        observation = self.cyber_env.reset()
+        return self.transform_observation(observation)
+
+    def transform_observation(self, observation) -> Observation:
+        # Flatten the action_mask field
```

Comment:
Should add TODOs here and elsewhere in this file to mention the observation space is currently the attacker's and needs to be changed to a custom defender view.